File: AndroidProject/CollectorAgentMind/src/main/java/com/example/collectoragentmind/codelets/behaviors/EatApple.java
package com.example.collectoragentmind.codelets.behaviors;

import com.example.collectoragentmind.CollectorAgentMind;
import com.example.collectoragentmind.util.Apple;
import com.example.collectoragentmind.util.Vector2;

import java.lang.reflect.Array;
import java.util.ArrayList;

import br.unicamp.cst.core.entities.Codelet;
import br.unicamp.cst.core.entities.Memory;
import br.unicamp.cst.core.entities.MemoryObject;

public class EatApple extends Codelet {

    /*
        For all apples in reach, as per the touchMO,
        adds its name in the handsMO
    */

    private MemoryObject touchMO;
    private MemoryObject handsMO;

    @Override
    public void accessMemoryObjects() {
        touchMO = (MemoryObject) this.getInput("TOUCH");
        handsMO = (MemoryObject) this.getOutput("HANDS");
    }

    @Override
    public void calculateActivation() {

    }

    private boolean appleInList(ArrayList<Apple> list, Apple apple) {
        for (int i = 0; i < list.size(); i ++) {
            if (list.get(i).name.equals(apple.name)) {
                return true;
            }
        }
        return false;
    }

    private int getApplePositionByName(ArrayList<Apple> list, String name) {
        for (int i = 0; i < list.size(); i++) {
            if (list.get(i).name.equals(name)) {
                return i;
            }
        }
        return -1;
    }

    @Override
    public void proc() {
        ArrayList<Apple> touchedApples = (ArrayList<Apple>) touchMO.getI();

        if (touchedApples.size() > 0) {
            ArrayList<Apple> applesInHand = (ArrayList<Apple>) handsMO.getI();

            for (int i = 0; i < touchedApples.size(); i++) {
                Apple newApple = touchedApples.get(i);

                if (!appleInList(applesInHand, newApple)) {
                    applesInHand.add(newApple);
                }
            }

            handsMO.setI(applesInHand);
        } else {
            handsMO.setI(new ArrayList<Apple>());
        }
    }
}

File: GodotKotlin/CoreModelTrail/src/main/kotlin/com/example/game/cst/codelets/behaviors/EatApple.kt
package com.example.game.cst.codelets.behaviors

import br.unicamp.cst.core.entities.Codelet
import br.unicamp.cst.core.entities.MemoryObject
import com.example.game.environment.Apple

class EatApple : Codelet() {
    /*
        For all apples in reach, as per the touchMO,
        adds its name in the handsMO
    */

    private var touchMO: MemoryObject? = null
    private var handsMO: MemoryObject? = null

    override fun accessMemoryObjects() {
        touchMO = getInput("TOUCH") as MemoryObject
        handsMO = getOutput("HANDS") as MemoryObject
    }

    override fun calculateActivation() {}

    private fun getApplePositionByCode(list: ArrayList<Apple>, name: String): Int {
        for (i in list.indices) {
            if (list[i].getCode().equals(name)) {
                return i
            }
        }
        return -1
    }

    override fun proc() {
        val touchedApples: ArrayList<Apple> = touchMO!!.i as ArrayList<Apple>
        if (touchedApples.size > 0) {
            var applesInHand: ArrayList<Apple> = handsMO!!.i as ArrayList<Apple>
            for (apple in touchedApples) {
                if (apple !in applesInHand) {
                    applesInHand.add(apple)
                }
            }
            handsMO?.i = applesInHand
        } else {
            handsMO?.i = ArrayList<Apple>()
        }
    }
}